reyna/src/com/b2msolutions/reyna/messageProvider/BatchProvider.java

```diff
@@ -0,0 +1,136 @@
+public class BatchProvider implements IMessageProvider {
+
+    private static final String TAG = "BatchProvider";
+
+    private static final String PeriodicBackoutCheckTAG = "BatchProvider";
+
+    private boolean batchDeleted = false;
+
+    protected BatchConfiguration batchConfiguration;
+
+    protected Repository repository;
+
+    protected PeriodicBackoutCheck periodicBackoutCheck;
+
+    protected Clock clock;
+
+    public BatchProvider(Context context) {
+        Logger.v(BatchProvider.TAG, "MessageProvider");
+
+        this.repository = new Repository(context);
+        this.batchConfiguration = new BatchConfiguration(context);
+        this.periodicBackoutCheck = new PeriodicBackoutCheck(context);
+        this.batchDeleted = false;
+        this.clock = new Clock();
+    }
+
+    public Message getNext() throws URISyntaxException {
+        Logger.v(BatchProvider.TAG, "getNext");
+
+        Message message = this.repository.getNext();
+        Batch batch = new Batch();
+
+        Header[] headers = null;
+        int count = 0;
+        long size= 0;
+        long maxMessagesCount = this.batchConfiguration.getBatchMessageCount();
+        long maxBatchSize = this.batchConfiguration.getBatchMessagesSize();
+
+        while (message != null && count <  maxMessagesCount && size < maxBatchSize) {
+            headers = message.getHeaders();
+            batch.add(message);
+
+            size = this.getSize(batch);
+            count++;
+            message = this.repository.getNext();
+        }
+
+        if (size > maxBatchSize) {
+            batch.removeLastMessage();
+        }
+
+        if (batch.getEvents().size() > 0) {
+            BatchMessage batchMessage = batch.getEvents().get(batch.getEvents().size() - 1);
+            URI uri = URI.create(batchMessage.getUrl());
+
+            String body = new Gson().toJson(batch, Batch.class);
+
+            return new Message(batchMessage.getReynaId(), this.getBatchUploadUrl(uri), body, headers);
+        }
+
+        return null;
+    }
+
+    public void delete(Message message) {
+        Logger.v(BatchProvider.TAG, "delete");
+
+        this.repository.deleteMessagesFrom(message.getId());
+        this.batchDeleted = true;
+    }
+
+    @Override
+    public void close() {
+        Logger.v(BatchProvider.TAG, "close");
+
+        if (batchDeleted) {
+            this.periodicBackoutCheck.record(PeriodicBackoutCheckTAG);
+        }
+
+        batchDeleted = false;
+        this.repository.close();
+    }
+
+    @Override
+    public boolean canSend() {
+        Logger.v(BatchProvider.TAG, "canSend");
+
+        if(this.periodicBackoutCheck.timeElapsed(PeriodicBackoutCheckTAG, this.batchConfiguration.getSubmitInterval())) {
+            Logger.v(BatchProvider.TAG, "canSend true, timeElapsed");
+            return true;
+        }
+
+        if(this.clock.isDifferentDay(this.clock.currentTimeMillis() + this.batchConfiguration.getCheckInterval(), this.clock.currentTimeMillis())) {
```

Comment:
isDifferentDay. to ensure that we are going to send messages at least once a day.
in case we have less than 100 message. we need to send these messages not wait until reach 100 messages.
@Youhana-Hana